### PR TITLE
Add implementation and tests for the JobRunrProDeadlineManager.

### PR DESCRIPTION
**Pull request by gklijs**
It needs the correct credentials and https://github.com/AxonFramework/AxonFramework/pull/2564 to be merged before it will build.

### CODE REVIEW

File: jobrunrpro/src/main/java/org/axonframework/extensions/jobrunrpro/scheduling/JobRunrProEventScheduler.java [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+        /**
+         * Sets the {@link String} used for creating the scheduled jobs in JobRunr. Defaults to
+         * {@code AxonScheduledEvent}.
+         *
+         * @param jobName a {@link JobScheduler} used for naming the job
+         * @return the current Builder instance, for fluent interfacing
+         */
+        public Builder jobName(String jobName) {
+            assertNonNull(jobName, "JobName may not be null");
```
Comment (smcvb) [inline jobrunrpro/src/main/java/org/axonframework/extensions/jobrunrpro/scheduling/JobRunrProEventScheduler.java:151]: I should've reacted on this in the AF PR too...but I think `BuilderUtils#assertNonEmpty` would be better suited for `String` validation in the builder.

File: jobrunrpro/src/main/java/org/axonframework/extensions/jobrunrpro/util/JobRunrProUtils.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import org.jobrunr.jobs.states.StateName;
+import org.jobrunr.scheduling.JobScheduler;
+import org.jobrunr.storage.JobSearchRequest;
+import org.jobrunr.storage.StorageProvider;
+
+import java.util.List;
+
+import static org.jobrunr.storage.JobSearchRequestBuilder.aJobSearchRequest;
+import static org.jobrunr.storage.PageRequest.ascOnUpdatedAt;
+
```
Comment (smcvb) [inline jobrunrpro/src/main/java/org/axonframework/extensions/jobrunrpro/util/JobRunrProUtils.java:29]: I'm missing JavaDoc throughout this class.

File: jobrunrpro/src/main/java/org/axonframework/extensions/jobrunrpro/util/JobRunrProUtils.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+public abstract class JobRunrProUtils {
```
Comment (smcvb) [inline jobrunrpro/src/main/java/org/axonframework/extensions/jobrunrpro/util/JobRunrProUtils.java:30]: I'm missing a test class for this.

File: jobrunrpro/src/test/java/org/axonframework/extensions/jobrunrpro/deadline/JobRunrProDeadlineManagerBuilderTest.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        assertThrows(AxonConfigurationException.class, () -> builder.transactionManager(null));
+    }
+
+    @Test
+    void whenSettingScopeAwareProviderWithNullThrowError() {
+        assertThrows(AxonConfigurationException.class, () -> builder.scopeAwareProvider(null));
+    }
+
+    @Test
+    void cancelImplemented() {
```
Comment (smcvb) [inline jobrunrpro/src/test/java/org/axonframework/extensions/jobrunrpro/deadline/JobRunrProDeadlineManagerBuilderTest.java:94]: Nit: Being a bit of an ass here, but the `cancelImplemented` and the `cancelAllWithinScopeImplemented` aren't tests I'd put under the "builder" group. So, I would expect this class to be called `JobRunrProDeadlineManagerTest` instead.

File: jobrunrpro/src/test/java/org/axonframework/extensions/jobrunrpro/scheduling/JobRunrProEventSchedulerBuilderTest.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        assertThrows(AxonConfigurationException.class, () -> builder.transactionManager(null));
+    }
+
+    @Test
+    void whenSettingEventBusWithNullThrowError() {
+        assertThrows(AxonConfigurationException.class, () -> builder.eventBus(null));
+    }
+
+    @Test
+    void cancelAllInvokesMock() {
```
Comment (smcvb) [inline jobrunrpro/src/test/java/org/axonframework/extensions/jobrunrpro/scheduling/JobRunrProEventSchedulerBuilderTest.java:93]: Nit: Being a bit of an ass here, but the `cancelAllInvokesMock` and the `cancelAllWithNameInvokesMock` aren't tests I'd put under the "builder" group. So, I would expect this class to be called `JobRunrProEventSchedulerTest` instead.

File: jobrunrpro/src/test/java/org/axonframework/extensions/jobrunrpro/scheduling/JobRunrProEventSchedulerBuilderTest.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        assertThrows(AxonConfigurationException.class, () -> builder.build());
+    }
+
+    @Test
+    void whenSettingSchedulerWithNullThrowError() {
+        assertThrows(AxonConfigurationException.class, () -> builder.jobScheduler(null));
+    }
+
+    @Test
+    void whenSettingstorageProviderWithNullThrowError() {
```
Comment (smcvb) [inline jobrunrpro/src/test/java/org/axonframework/extensions/jobrunrpro/scheduling/JobRunrProEventSchedulerBuilderTest.java:78]:
```suggestion
    void whenSettingStorageProviderWithNullThrowError() {
```

File: jobrunrpro/src/test/java/org/axonframework/extensions/jobrunrpro/deadline/JobRunrProDeadlineManagerBuilderTest.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        assertThrows(AxonConfigurationException.class, () -> builder.build());
+    }
+
+    @Test
+    void whenSettingSchedulerWithNullThrowError() {
+        assertThrows(AxonConfigurationException.class, () -> builder.jobScheduler(null));
+    }
+
+    @Test
+    void whenSettingstorageProviderWithNullThrowError() {
```
Comment (smcvb) [inline jobrunrpro/src/test/java/org/axonframework/extensions/jobrunrpro/deadline/JobRunrProDeadlineManagerBuilderTest.java:79]:
```suggestion
    void whenSettingStorageProviderWithNullThrowError() {
```

**Review by smcvb — CHANGES_REQUESTED** [review]
Mainly small fry, but one pointer that's worth discussing: do we really need a `JobRunrProEventScheduler`?

File: jobrunrpro/src/main/java/org/axonframework/extensions/jobrunrpro/deadline/JobRunrProDeadlineManager.java [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+
+    @Override
+    public void cancelAllWithinScope(@Nonnull String deadlineName, @Nonnull ScopeDescriptor scope) {
+        Span span = spanFactory.createInternalSpan(
+                () -> getSpanClassName() + ".cancelAllWithinScope(" + deadlineName + ")");
+        runOnPrepareCommitOrNow(span.wrapRunnable(() -> deleteAll(getCombinedLabel(serializer, deadlineName, scope))));
+    }
+
+    private void deleteAll(String label) {
+        deleteAllPendingJobsByLabel(jobScheduler, storageProvider, label, DELETE_REASON);
```
[the remark this answers is not in the document]
Comment (gklijs) [reply jobrunrpro/src/main/java/org/axonframework/extensions/jobrunrpro/deadline/JobRunrProDeadlineManager.java:115]: It's the one from the `JobRunrDeadlineManager`, as intended.

File: README.md [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+This repository provides an extension to the Axon Framework: JobRunr Pro.
+It provides functionality to leverage JobRunr to store and execute jobs. In particular and implementation for both an
+event scheduler and a deadline manager, that allows deleting jobs by their label.
+
+For more information on anything Axon, please visit our website, [http://axoniq.io](http://axoniq.io).
+
+## Getting started
+
+The [reference guide](https://docs.axoniq.io) contains a separate chapter for all the extensions.
+The jobRunr Pro extension description can be found [here](https://docs.axoniq.io/reference-guide/extensions/jobrunrpro).
```
[the remark this answers is not in the document]
Comment (gklijs) [reply README.md:26]: added, https://github.com/AxonIQ/reference-guide/issues/310

File: jobrunrpro/src/main/java/org/axonframework/extensions/jobrunrpro/scheduling/JobRunrProEventScheduler.java [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+        this.jobScheduler = builder.jobScheduler;
+        this.storageProvider = builder.storageProvider;
+        this.jobName = builder.jobName;
+    }
+
+    /**
+     * Cancels all the events scheduled with the current {@code jobname}. This method has no impact on events which have
+     * already been triggered.
+     */
+    public void cancelAll() {
```
[the remark this answers is not in the document]
Comment (gklijs) [reply jobrunrpro/src/main/java/org/axonframework/extensions/jobrunrpro/scheduling/JobRunrProEventScheduler.java]: No longer relevant, with the JobRuneEventScheduler being removed.

File: jobrunrpro/src/main/java/org/axonframework/extensions/jobrunrpro/scheduling/JobRunrProEventScheduler.java [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+            assertNonNull(jobScheduler, "JobScheduler may not be null");
+            this.jobScheduler = jobScheduler;
+            return this;
+        }
+
+        /**
+         * Sets the {@link StorageProvider} used for the cancel all methods. Those will return an error when called
+         * while the storage provider has not been set.
+         *
+         * @param storageProvider a {@link StorageProvider} used for scheduling and triggering purposes of the
```
[the remark this answers is not in the document]
Comment (gklijs) [reply jobrunrpro/src/main/java/org/axonframework/extensions/jobrunrpro/scheduling/JobRunrProEventScheduler.java]: No longer relevant, with the JobRuneEventScheduler being removed.

File: jobrunrpro/src/main/java/org/axonframework/extensions/jobrunrpro/scheduling/JobRunrProEventScheduler.java (hunk 1 above)
Comment (gklijs) [reply jobrunrpro/src/main/java/org/axonframework/extensions/jobrunrpro/scheduling/JobRunrProEventScheduler.java]: No longer relevant, with the JobRuneEventScheduler being removed.

File: jobrunrpro/src/test/java/org/axonframework/extensions/jobrunrpro/scheduling/JobRunrProEventSchedulerBuilderTest.java (hunk 5 above)
Comment (gklijs) [reply jobrunrpro/src/test/java/org/axonframework/extensions/jobrunrpro/scheduling/JobRunrProEventSchedulerBuilderTest.java]: No longer relevant.

File: README.md [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+As such it provides you the necessary building blocks to follow these principles.
+Building blocks like Aggregate factories and Repositories, Command, Event and Query Buses and an Event Store.
+The framework provides sensible defaults for all of these components out of the box.
+
+This set up helps you create a well structured application without having to bother with the infrastructure.
+The main focus can thus become your business functionality.
+
+This repository provides an extension to the Axon Framework: JobRunr Pro.
+It provides functionality to leverage JobRunr to store and execute jobs. In particular and implementation for both an
```
[the remark this answers is not in the document]
Comment (gklijs) [reply README.md]: added

File: jobrunrpro/src/main/java/org/axonframework/extensions/jobrunrpro/deadline/JobRunrProDeadlineManager.java [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+            assertNonNull(jobScheduler, "JobScheduler may not be null");
+            this.jobScheduler = jobScheduler;
+            return this;
+        }
+
+        /**
+         * Sets the {@link StorageProvider} used for the cancel all methods. Those will return an error when called
+         * while the storage provider has not been set.
+         *
+         * @param storageProvider a {@link StorageProvider} used for scheduling and triggering purposes of the
```
[the remark this answers is not in the document]
Comment (gklijs) [reply jobrunrpro/src/main/java/org/axonframework/extensions/jobrunrpro/deadline/JobRunrProDeadlineManager.java]: changed

File: jobrunrpro/src/main/java/org/axonframework/extensions/jobrunrpro/util/JobRunrProUtils.java (hunk 2 above)
Comment (gklijs) [reply jobrunrpro/src/main/java/org/axonframework/extensions/jobrunrpro/util/JobRunrProUtils.java:29]: Added

File: jobrunrpro/src/main/java/org/axonframework/extensions/jobrunrpro/util/JobRunrProUtils.java (hunk 3 above)
Comment (gklijs) [reply jobrunrpro/src/main/java/org/axonframework/extensions/jobrunrpro/util/JobRunrProUtils.java:36]: Added tests

File: jobrunrpro/src/test/java/org/axonframework/extensions/jobrunrpro/deadline/JobRunrProDeadlineManagerBuilderTest.java (hunk 4 above)
Comment (gklijs) [reply jobrunrpro/src/test/java/org/axonframework/extensions/jobrunrpro/deadline/JobRunrProDeadlineManagerBuilderTest.java]: moved

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+   the [AxonIQ's discussion platform](https://discuss.axoniq.io/).
+
+## Code Contributions
+
+If you're contributing code, please take care of the following:
+
+### Contributor Licence Agreement
+
+To keep everyone out of trouble (both you and us), we require that all contributors (digitally) sign a Contributor
+License Agreement. Basically, the agreement says that we may freely use the code you contribute to the Axon Framework
```
Comment (Morlack) [inline CONTRIBUTING.md:24]: Kafka extension is mentioned here

File: jobrunrpro-spring-boot-3-integrationtests/pom.xml [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
+
+    <dependencyManagement>
+        <dependencies>
+            <dependency>
+                <groupId>org.testcontainers</groupId>
+                <artifactId>testcontainers-bom</artifactId>
+                <version>1.17.6</version>
```
Comment (Morlack) [inline jobrunrpro-spring-boot-3-integrationtests/pom.xml:118]: This version is duplicated, in bom and dependency itself. Perhaps move it to a property?

**Review by Morlack — APPROVED** [review]
2 small comments, otherwise looks good to me :+1:

File: CONTRIBUTING.md (hunk 14 above)
Comment (gklijs) [reply CONTRIBUTING.md:24]: Replaced by JobRunr Pro, good catch.

File: jobrunrpro-spring-boot-3-integrationtests/pom.xml (hunk 15 above)
Comment (gklijs) [reply jobrunrpro-spring-boot-3-integrationtests/pom.xml]: Using property version now.

File: README.md [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+This repository provides an extension to the Axon Framework: JobRunr Pro.
+It provides functionality to leverage [JobRunr](https://www.jobrunr.io/) to store and execute jobs. In particular and
+implementation for a deadline manager, that allows deleting jobs by their label.
+
+For more information on anything Axon, please visit our website, [http://axoniq.io](http://axoniq.io).
+
+## Getting started
+
+The [reference guide](https://docs.axoniq.io) contains a separate chapter for all the extensions.
+The jobRunr Pro extension description can be found [here](https://docs.axoniq.io/reference-guide/extensions/jobrunrpro).
```
Comment (smcvb) [inline README.md:26]:
```suggestion
The JobRunr Pro extension description can be found [here](https://docs.axoniq.io/reference-guide/extensions/jobrunrpro).
```

File: jobrunrpro-spring-boot-autoconfigure/src/test/java/org/axonframework/extensions/jobrunrpro/autoconfig/AxonJobRunrProAutoConfigurationTest.java [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+import org.springframework.jmx.support.RegistrationPolicy;
+import org.springframework.test.context.ContextConfiguration;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.Mockito.*;
+
+class AxonJobRunrProAutoConfigurationTest {
+
+    @Test
+    void eventSchedulerAndDeadlineManagercreated() {
```
Comment (smcvb) [inline jobrunrpro-spring-boot-autoconfigure/src/test/java/org/axonframework/extensions/jobrunrpro/autoconfig/AxonJobRunrProAutoConfigurationTest.java:37]:
```suggestion
    void jobRunrDeadlineManagerIsCreated() {
```

File: README.md [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+based on the principles of Domain Driven Design, Command-Query Responsibility Segregation (CQRS) and Event Sourcing.
+
+As such it provides you the necessary building blocks to follow these principles.
+Building blocks like Aggregate factories and Repositories, Command, Event and Query Buses and an Event Store.
+The framework provides sensible defaults for all of these components out of the box.
+
+This set up helps you create a well structured application without having to bother with the infrastructure.
+The main focus can thus become your business functionality.
+
+This repository provides an extension to the Axon Framework: JobRunr Pro.
```
Comment (smcvb) [inline README.md:17]: I may be missing it, but I do not see the link to the website yet.
So, what do you think, would this make sense:

```suggestion
This repository provides an extension to the Axon Framework: [JobRunr Pro](https://www.jobrunr.io/en/pricing/).
```

**Review by smcvb — APPROVED** [review]
Just some small fry remaining. Other than that, I think this looks good :+1:

**Pull request merged**